lib2/client/okta.go

```
@@ -0,0 +1,667 @@
+func (o *OktaClient) getSessionCookieKeyringKey() string {
+	return "okta-session-cookie-" + o.creds.Username + "-" + o.creds.Domain
+}
+
+// Takes a session cookie in the cookie jar and saves it in the keychain,
+// this allows it to be used across invocations where the client making the
+// request is destroyed/created between requests.
+func (o *OktaClient) SaveSessionCookie() (err error) {
+
+	if o.Keyring == nil {
+		return fmt.Errorf("Session NOT saved. Reason: Session Backend not defined")
+	}
+	cookies := o.CookieJar.Cookies(o.BaseURL)
+	for _, cookie := range cookies {
+		if cookie.Name == "sid" {
+			newCookieItem := keyring.Item{
+				Key:                         o.getSessionCookieKeyringKey(),
+				Data:                        []byte(cookie.Value),
+				Label:                       "okta session cookie for " + o.creds.Username,
+				KeychainNotTrustApplication: false,
+			}
+			err = (*o.Keyring).Set(newCookieItem)
+			if err != nil {
+				return
+			}
+			log.Debug("Saving Okta Session Cookie: ", cookie.Value)
+		}
+	}
+	return
+}
+
+// Sends a request to the Okta Sessions API to validate if the session cookie
+// is valid or not. This doesn't always mean that the session can be used for
+// all Okta applications but it does accurately fetch the state of the session.
+func (o *OktaClient) ValidateSession() (sessionValid bool, err error) {
+	var mySessionResponse *http.Response
+	sessionValid = false
+
+	log.Debug("Checking if we have a valid Okta session")
+	mySessionResponse, err = o.Request("GET", "api/v1/sessions/me", url.Values{}, []byte{}, "json", false)
+	if err != nil {
+		return
+	}
+	defer mySessionResponse.Body.Close()
+
+	// https://developer.okta.com/docs/reference/api/sessions/#get-current-session
+	// "if the session is invalid, a 404 Not Found response will be returned."
+	// checking for ok status (200) is adequate to see if the session is still valid.
+	sessionValid = mySessionResponse.StatusCode == http.StatusOK
```

I wonder if this should just return an error if the session is invalid, instead of also having this bool

---

I wanted to have a separate `bool` so that we can determine if the request is failing to complete vs the session not being valid. Not sure how important that is but that is why I split it out.

---

I'd say typed errors are a better way to deal with that (along with `errors.Is/As`), but I can see that having a `bool` might be simpler.